---
title: "Initial Survival"
author: "Shun Sambongi"
output: html_notebook
---

## Setup

### Load required packages

```{r}
source('shun.R')
use.package(dplyr)
use.package(magrittr)
use.package(survival)
use.package(survminer)
use.package(tidyr)
```

### Import data

```{r}
phenotype <- import.data("../Data/phenotype_filtered.tsv")
mutations <- import.data("../Data/mutations_filtered.tsv")
copy_nums <- import.data("../Data/copy_nums_filtered.tsv")
```

### Format data

```{r}
# Phenotype data
phenotype.filtered <- phenotype %>% 
  select(sampleID, days_to_death, days_to_last_followup, vital_status) %>%
  rename(sample.id=sampleID) %>%
  mutate(years_to_death=days_to_death/365, 
         years_to_last_followup=days_to_last_followup/365) %T>% print()

# Mutation data
mutations.tidy <- mutations %>% 
  rename(gene.symbol=sample) %>% 
  gather(sample.id, mutation.status, -gene.symbol) %T>% print()

# Copy number data
copy_nums.tidy <- copy_nums %>%
  rename(gene.symbol=`Gene Symbol`) %>% 
  gather(sample.id, cnv.status, -gene.symbol) %>% 
  mutate(cnv.status=floor(abs(cnv.status) / 2))
```


## Analysis

### Survival: TP53 mutation

```{r}
mutations.survival <- phenotype.filtered %>% 
  mutate(time1=ifelse(is.na(years_to_death), 
                      years_to_last_followup, 
                      years_to_death), 
         event=as.numeric(vital_status=="DECEASED")) %>% 
  select(sample.id, time1, event) %>%
  inner_join(mutations.tidy %>% filter(gene.symbol=="TP53")) %>%
  select(sample.id, time1, event, mutation.status) %T>% print()

mutations.fit <- survfit(Surv(time1, 
                              event=event, 
                              type="right") ~ mutation.status,
                         data=mutations.survival)
ggsurvplot(mutations.fit, 
           legend.title="TP53 Mutation Status",
           legend.labs=c("No Mutation", "Mutation"),
           legend="right",
           xlab="Years after diagnosis",
           pval=TRUE)

save.figure("survival_prob_TP53_mutation.png")
```

### Survival: 8q CNV

Get chromosome regions using biomaRt

```{r}
use.package(biomaRt, from=bioc)
ensembl <- useMart("ensembl", dataset="hsapiens_gene_ensembl")

# Only use standard human chromosomes
normal.chroms <- c(1:22, "X", "Y", "M")

# Filter on HGNC symbol and chromosome, retrieve genomic location and band
gene.symbols <- copy_nums.tidy$gene.symbol %>% unique()

chrom.regions <- getBM(attributes=c("hgnc_symbol", "chromosome_name", "band"),
                       filters=c("hgnc_symbol", "chromosome_name"),
                       values=list(hgnc_symbol=gene.symbols, 
                                   chromosome_name=normal.chroms), 
                       mart=ensembl)
detach.package(biomaRt)
head(chrom.regions)
```

Parse out general chromosome regions

```{r}
regions <- chrom.regions %>% 
  rename(gene.symbol=hgnc_symbol, chrom=chromosome_name) %>% 
  mutate(arm=sub("^([pq])\\d+(\\.\\d+)?", "\\1", band)) %>%
  mutate(chrom.region=paste0(chrom, arm)) %T>% print() %>%
  select(gene.symbol, chrom.region)
```

Plot KM curve

```{r}
cnv.survival <- copy_nums.tidy %>% 
  inner_join(regions) %>%
  filter(chrom.region=="8q") %>% 
  group_by(sample.id) %>%
  summarise(cnv.status=sum(cnv.status)>0) %>%
  inner_join(phenotype.filtered) %>%
  mutate(time1=ifelse(is.na(years_to_death), 
                      years_to_last_followup, 
                      years_to_death), 
         event=as.numeric(vital_status=="DECEASED")) %>% 
  select(sample.id, time1, event, cnv.status) %T>% print()
  
cnv.fit <-survfit(Surv(time1, event=event, type="right") ~ cnv.status, data=cnv.survival)

ggsurvplot(cnv.fit, 
           legend.title="8q CNV Status",
           legend.labs=c("No CNV", "CNV"),
           legend="right",
           xlab="Years after diagnosis",
           pval=T)

save.figure("survival_prob_8q_cnv.png")
```

### Survival: TP53 mutation and 8q CNV

```{r}
tp53.8q.survival <- mutations.survival %>%
  inner_join(cnv.survival) %>%
  mutate(cnv.status=ifelse(cnv.status, 2, 0),
         tp53.8q=mutation.status + cnv.status,
         tp53.8q=recode(tp53.8q, 
                        `0`="Normal", 
                        `1`="TP53", 
                        `2`="8q", 
                        `3`="TP53-8q")) %>%
  select(sample.id, time1, event, tp53.8q) %T>% print()
  
tp53.8q.fit <- survfit(Surv(time1, event=event, type="right") ~ tp53.8q,
                       data=tp53.8q.survival)

ggsurvplot(tp53.8q.fit, 
           legend.title="Genomic Events",
           legend="right",
           legend.labs=sub("^tp53\\.8q=", "", names(tp53.8q.fit$strata)),
           xlab="Years after diagnosis", pval=T)

save.figure("survival_prob_TP53-8q_event.png")
```

### Cox Proportional Hazard's Model

```{r}
fit <- coxph(Surv(time1, event) ~ mutation.status, data=mutations.survival)
summary(fit)

ggcoxadjustedcurves(fit)
```

### Misc

```{r}
phenotype %>% select(days_to_new_tumor_event_after_initial_treatment) %>% na.omit() %>% nrow()
```


### Survival:Censored at 5 years

Censoring event time at 5 years. Any person who died after 5 years is considered "alive" and their time of event is set to 5 years. 

```{r}
## Are we censoring correctly?
mutations.survival.censored <- mutations.survival %>% mutate(time=ifelse(time1 > 5, 5, time1),
                                                             event=ifelse(time1 > 5, 0, event)) 
mutations.fit.censored <- survfit(Surv(time, 
                                       event=event, 
                                       type="right") ~ mutation.status,
                                  data=mutations.survival.censored)
ggsurvplot(mutations.fit.censored, 
           legend.title="TP53 Mutation Status",
           legend.labs=c("No Mutation", "Mutation"),
           legend="right",
           xlab="Years after diagnosis",
           pval=TRUE)

save.figure("tp53_KM_censored.png")


print(survfit(Surv(time, event=event, type="right") ~ 1, data=mutations.survival.censored), rmean="individual")
```


```{r}
cnv.survival.censored <- cnv.survival %>% mutate(time=ifelse(time1 > 5, 5, time1),
                                                 event=ifelse(time1 > 5, 0, 1)) 

cnv.fit.censored <-survfit(Surv(time, event=event, type="right") ~ cnv.status,
                           data=cnv.survival.censored)

ggsurvplot(cnv.fit.censored, 
           legend.title="8q CNV Status",
           legend.labs=c("No CNV", "CNV"),
           legend="right",
           xlab="Years after diagnosis",
           pval=T)
```


```{r}
tp53.8q.survival.censored <- mutations.survival.censored %>%
  inner_join(cnv.survival.censored) %>%
  mutate(cnv.status=ifelse(cnv.status, 2, 0),
         tp53.8q=mutation.status + cnv.status,
         tp53.8q=recode(tp53.8q, 
                        `0`="Normal", 
                        `1`="TP53", 
                        `2`="8q", 
                        `3`="TP53-8q")) %>%
  select(sample.id, time, event, tp53.8q) %T>% print()
  
tp53.8q.fit.censored <- survfit(Surv(time, event=event, type="right") ~ tp53.8q,
                       data=tp53.8q.survival.censored)

ggsurvplot(tp53.8q.fit.censored, 
           legend.title="Genomic Events",
           legend="right",
           legend.labs=sub("^tp53\\.8q=", "", names(tp53.8q.fit$strata)),
           xlab="Years after diagnosis", pval=T)

```


### KM: Cancer types

```{r}
subtypes.surv <- phenotype %>% 
  select(PAM50Call_RNAseq,
         days_to_death,
         days_to_last_followup, 
         vital_status) %>% 
  mutate(time=ifelse(is.na(days_to_death), 
                     days_to_last_followup, 
                     days_to_death),
         event=ifelse(vital_status=="DECEASED", 1, 0),
         time=time/365.25,
         event=ifelse(time > 5, 0, event),
         time=ifelse(time > 5, 5, time)) %>% 
  replace_na(list(PAM50Call_RNAseq="Other")) %>% 
  select(subtype=PAM50Call_RNAseq, time, event) %T>%
  print()

subtypes.fit <- survfit(Surv(time, event) ~ subtype, data=subtypes.surv)

ggsurvplot(subtypes.fit, legend="right")
```


